File: statcruncher/api/views/leaderboard.py
import logging

from rest_framework import status
from rest_framework.views import APIView
from rest_framework.response import Response

from django.views.decorators.cache import cache_page
from django.utils.decorators import method_decorator

from ..apps import STATCRUNCHER

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)

class Leaderboard():

	@staticmethod
	def get_leaderboard_response(period, requested_stats):
		if not requested_stats:
			return Response({'error': 'No stats requested'}, status=status.HTTP_400_BAD_REQUEST)
		if requested_stats not in ['gamesPlayed', 'gamesWon', 'gamesLost']:
			return Response({'error': 'Invalid stats requested'}, status=status.HTTP_400_BAD_REQUEST)

		leaderboard = STATCRUNCHER.get_leaderboard(period, requested_stats)
		return Response(leaderboard, status=status.HTTP_200_OK)

	class Daily(APIView):

		@method_decorator(cache_page(60 * 2, key_prefix='leaderboard'))
		def get(self, request, *args, **kwargs):
			requested_stats = request.query_params.get('stats', None)

			return Leaderboard.get_leaderboard_response("daily", requested_stats)

	class Weekly(APIView):

		@method_decorator(cache_page(60 * 2, key_prefix='leaderboard'))
		def get(self, request, *args, **kwargs):
			requested_stats = request.query_params.get('stats', None)

			return Leaderboard.get_leaderboard_response("weekly", requested_stats)

	class Lifetime(APIView):

		@method_decorator(cache_page(60 * 2, key_prefix='leaderboard'))
		def get(self, request, *args, **kwargs):
			requested_stats = request.query_params.get('stats', None)

			return Leaderboard.get_leaderboard_response("lifetime", requested_stats)

File: frontend/src/api/leaderboard.js
import API from './api';

/**
 * Fetch leaderboard based on time frame (daily, weekly, lifetime) and stats (gamesPlayed, gamesWon, gamesLost)
 * @param {string} timeFrame - daily, weekly, lifetime
 * @param {string} stats - gamesPlayed, gamesWon, gamesLost
 * @returns leaderboard data
 */
export const getLeaderboard = async (timeFrame, stats) => {
	try {
		// API call using the timeFrame and stats parameters
		const res = await API.get(`leaderboards/${timeFrame}?stats=${stats}`);
		return res.data;
	} catch (err) {
		console.error(err.response?.data?.error || 'An error occurred');
		return [];
	}
};
